### DISCUSSION TITLE
Creating ID3D11Texture2D from Windows.Graphics.DirectX.Direct3D11.IDirect3DSurface

### DISCUSSION THREAD

**Discussion by triandco** [dotnet/Silk.NET#1699] [General] [answered]
I'm trying to use Silk.NET in place of [SharpDX](https://github.com/sharpdx/SharpDX) to capture the screen as video. I'm following [Microsoft tutorial](https://learn.microsoft.com/en-us/windows/uwp/audio-video-camera/screen-capture-video) to do this. However, I'm stuck on a generating a Texture2D from a Windows.Graphics.DirectX.Direct3D11.IDirect3DSurface

Here's the SharpDX code to do so as provided by Microsoft which runs fine.
```csharp
internal static Guid ID3D11Texture2D = new Guid("6f15aaf2-d208-4e89-9ab4-489535d34f9c");
[Guid("A9B3D012-3DF2-4EE3-B8D1-8695F457D3C1")]
[InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
[ComVisible(true)]
interface IDirect3DDxgiInterfaceAccess
{
    IntPtr GetInterface([In] ref Guid iid);
};

public static class Helper(){
  internal static SharpDX.Direct3D11.Texture2D CreateSharpDXTexture2D(IDirect3DSurface surface)
    {
        var access = (IDirect3DDxgiInterfaceAccess)surface;
        var d3dPointer = access.GetInterface(ID3D11Texture2D);
        var d3dSurface = new SharpDX.Direct3D11.Texture2D(d3dPointer);
        return d3dSurface;
    }

}
```

My attempt to translate it to Silk.NET is as follow:
```csharp
[Guid("A9B3D012-3DF2-4EE3-B8D1-8695F457D3C1")]
[InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
[ComVisible(true)]
interface IDirect3DDxgiInterfaceAccess
{
    IntPtr GetInterface([In] ref Guid iid);
};
public static class Helper(){
  public static ComPtr<ID3D11Texture2D> Texture2DFromSurface(IDirect3DSurface surface)
  {
      var access = surface.As<IDirect3DDxgiInterfaceAccess>();
      IntPtr pUnknown = access.GetInterface(ID3D11Texture2D.Guid);
      var t = new ID3D11Texture2D((void**)pUnknown.ToPointer());
      return &t;
  }
}
```

this code however results in an exception.
```
System.AccessViolationException: 'Attempted to read or write protected memory. This is often an indication that other memory is corrupt.'
```

I haven't work with CSharp at this level before 😅 so I'm sure if this is the right way to turn the ```IntPtr``` into ```ID3D11Texture2D```. Any help would be much appreciated.

**Comment by Perksey** [comment]
Fun use case! I'll try and get IDirect3DDxgiInterfaceAccess exposed for Tuesday's update (and get you a working copy of the code :))

**Reply by triandco** [reply]
Thank you @Perksey!

I've been poking around a bit comparing SharpDX and Silk.NET implementation and the following code seems to work okay. However, I'm not quite sure if it is the right way to do it or if there is any consequent of doing this 😅.

Exposing IDirect3DDxgiInterfaceAccess with working copy would be hugely appreciated. 🙇‍♂️🙇‍♂️

```csharp
public static ComPtr<ID3D11Texture2D> Texture2DFromSurface(IDirect3DSurface surface)
{
  var access = surface.As<IDirect3DDxgiInterfaceAccess>();
  IntPtr pUnknown = access.GetInterface(ID3D11Texture2D.Guid);
  return (ID3D11Texture2D*)pUnknown;
}
```

**Reply by Perksey** [reply, marked as answer]
I've had a look at this, and will add IDirect3DDxgiInterfaceAccess in 2.18. However, I won't add any other WinRT APIs like GraphicsCapture primarily because 2.X isn't currently in a fit state for WinRT. I recommend using those APIs through CsWinRT for now.

Your error is here:
```cs
      var t = new ID3D11Texture2D((void**)pUnknown.ToPointer());
      return &t;
```

You should do
```cs
new ComPtr<ID3D11Texture2D>((ID3D11Texture2D*)access.GetInterface(ID3D11Texture2D.Guid)); 
```
instead. `ID3D11Texture2D` in Silk.NET is a structure wrapping a `void**`.

Note that this will also add a ref so watch out for memory leaks!
